File: day20/day20.cpp
#include <fstream>
#include <iostream>
#include <regex>
#include <unordered_map>
#include <vector>

#include <fmt/format.h>

using namespace std;

struct Vec
{
	int v[3];

	Vec& operator+=(const Vec& o) noexcept
	{
		v[0] += o.v[0];
		v[1] += o.v[1];
		v[2] += o.v[2];
		return *this;
	}

	bool operator==(const Vec& o) const noexcept
	{
		return v[0] == o.v[0] && v[1] == o.v[1] && v[2] == o.v[2];
	}

	size_t manhattan() const noexcept
	{
		return (v[0]>0?v[0]:-v[0]) +
			(v[1]>0?v[1]:-v[1]) +
			(v[2]>0?v[2]:-v[2]);
	}
};

namespace std
{
	template<>
	struct hash<Vec>
	{
		size_t operator()(const Vec& v) const noexcept
		{
			return (hash<int>{}(v.v[0]) << 32)
				^ (hash<int>{}(v.v[1]) << 16)
				^ (hash<int>{}(v.v[2]) << 0);
		}
	};
}

struct Particle
{
	size_t n;
	Vec p;
	Vec v;
	Vec a;

	void update() noexcept
	{
		v += a;
		p += v;
	}

	bool stable(int axis) const noexcept
	{
		return (a.v[axis] == 0 || (v.v[axis] ^ a.v[axis]) > 0)
			&& (v.v[axis] == 0 || (p.v[axis] ^ v.v[axis]) > 0);
	}
};

struct World
{
	vector<Particle> parts;

	void update(bool collide)
	{
		unordered_map<Vec, int> coll;
		for (auto& p: parts)
		{
			p.update();
			if (collide)
			{
				coll[p.p] += 1;
			}
		}
		if (collide)
		{
			parts.erase(
				remove_if(parts.begin(), parts.end(),
					  [&coll](const auto& p){
						  return coll[p.p] > 1;
					  }),
				parts.end()
				);
		}
	}

	bool stable(int axis) const
	{
		return all_of(
			parts.begin(),
			parts.end(),
			[axis](const auto& p){
				return p.stable(axis);
			});
	}
};

static size_t part1(World w)
{
	for (int axis = 0; axis < 3; axis++)
	{
		while (!w.stable(axis))
		{
			w.update(false);
		}
	}
	w.update(false);
	auto it = min_element(w.parts.begin(), w.parts.end(), [](const auto& a, const auto& b) {
		return a.p.manhattan() < b.p.manhattan();
	});
	return it->n;
}

static size_t part2(World w)
{
	for (int axis = 0; axis < 3; axis++)
	{
		while (!w.stable(axis))
		{
			w.update(true);
		}
	}
	w.update(true);
	return w.parts.size();
}

istream& operator>>(istream& input, World& w)
{
	static const regex pat("p=<\\s*(-?\\d+),(-?\\d+),(-?\\d+)>, v=<\\s*(-?\\d+),(-?\\d+),(-?\\d+)>, a=<\\s*(-?\\d+),(-?\\d+),(-?\\d+)>");

	string line;
	while (getline(input, line))
	{
		smatch sm;
		if (!regex_match(line, sm, pat))
		{
			continue;
		}
		Vec p{stoi(sm[1]), stoi(sm[2]), stoi(sm[3])};
		Vec v{stoi(sm[4]), stoi(sm[5]), stoi(sm[6])};
		Vec a{stoi(sm[7]), stoi(sm[8]), stoi(sm[9])};
		w.parts.emplace_back(Particle{w.parts.size(), p, v, a});
	}
	return input;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	World w;
	input>>w;
	input.close();
	if (w.parts.size() == 0)
	{
		cerr << "Cannot parse the data" << endl;
		return 1;
	}
	cout << fmt::format("Part1: {}", part1(w)) << endl
	     << fmt::format("Part2: {}", part2(w)) << endl;
	return 0;
}

File: day20/day20.py
#!/usr/bin/env python3
import re
import sys

def add(a, b):
    return (a[0]+b[0], a[1]+b[1], a[2]+b[2])

def distance(a):
    return sum(map(abs, a))

class Particle(object):
    def __init__(self, n, p, v, a):
        self.n = n
        self.p = p
        self.v = v
        self.a = a

    def update(self):
        self.v = add(self.v, self.a)
        self.p = add(self.p, self.v)

    def stable(self, axis):
        return self.v[axis] * self.a[axis] >= 0 and self.v[axis] * self.p[axis] >= 0

    def __repr__(self):
        return "{}: p={}, v={}, a={}".format(self.n, self.p, self.v, self.a)

def update(particles, collision):
    # update the world
    coll = {}
    for p in particles:
        p.update()
        coll[p.p] = coll.get(p.p, 0) + 1

    if collision:
        return [p for p in particles if coll[p.p] == 1]
    else:
        return particles

def part1(parts):
    for axis in range(3):
        while not all(p.stable(axis) for p in parts):
            parts = update(parts, False)

    parts = update(parts, False)
    p = min(parts, key = lambda x: distance(x.p))
    return p.n

def part2(parts):
    for axis in range(3):
        while not all(p.stable(axis) for p in parts):
            parts = update(parts, True)

    parts = update(parts, True)
    return len(parts)

def parse(txt):
    parts = []
    pat = re.compile(r"[pva]=<\s*(-?\d+),(-?\d+),(-?\d+)>")
    for row in txt.splitlines():
        lst = []
        for c in row.split(", "):
            m = pat.match(c)
            if m:
                lst.append(tuple(map(int, m.groups())))
        parts.append(Particle(len(parts), *lst))
    return parts

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    p = parse(txt)
    print("Part1:", part1(p))
    p = parse(txt)
    print("Part2:", part2(p))
